velociraptor/autoplotter/objects.py

```diff
@@ -745,20 +764,46 @@ def get_quantity_from_catalogue_with_mask(
             ).astype(bool)
 
             if self.select_structure_type is not None:
+                if self.select_structure_type == self.exclude_structure_type:
+                    raise AutoPlotterError(
+                        f"Cannot simultaneously select and exclude structure"
+                        " type {self.select_structure_type}"
+                    )
                 self.structure_mask = logical_and(
                     self.structure_mask,
                     catalogue.structure_type.structuretype
                     == self.select_structure_type,
                 )
 
+            elif self.exclude_structure_type is not None:
+                self.structure_mask = logical_and(
+                    self.structure_mask,
+                    catalogue.structure_type.structuretype
+                    != self.exclude_structure_type,
+                )
+                
             x = x[self.structure_mask]
             x.name = name
         elif self.select_structure_type is not None:
+            if self.select_structure_type == self.exclude_structure_type:
+                raise AutoPlotterError(
+                    f"Cannot simultaneously select and exclude structure"
+                    " type {self.select_structure_type}"
+                )
+            
             # Need to create mask
             self.structure_mask = (
                 catalogue.structure_type.structuretype == self.select_structure_type
             )
 
+            x = x[self.structure_mask]
+            x.name = name
+        elif self.exclude_structure_type is not None:
+            # Need to create mask
+            self.structure_mask = (
+                catalogue.structure_type.structuretype != self.exclude_structure_type
+            )
+
```

Comment:
```suggestion
            select_mask = np._s[:]
            exclude_mask = np._s[:]

            if self.select_structure_type is not None:
                if self.exclude_structure_type == self.select_structure_type:
                    raise AutoPlotterError(
                        f"Cannot simultaneously select and exclude structure"
                        " type {self.select_structure_type}"
                    )

                select_mask = catalogue.structure_type.structuretype == self.select_structure_type
                
            if self.exclude_structure_type is not None:
                exclude_mask =  catalogue.structure_type.structuretype != self.exclude_structure_type
                
            self.structure_mask = np.logical_and.reduce(
                [
                    self.structure_mask,
                    select_mask,
                    exclude_mask,
                ]
            )

            x = x[self.structure_mask]
            x.name = name
```

Reply:
this looks more compact but doesn't work currently as far as I can tell. I figured that you meant `s_` as this exists in numpy, and explicitly import it as you do for other python modules, but beyond that unfamiliar with this approach

Reply:
Ah I see we will need to import those functions, yeah.

`s_` is basically a shortcut for turning fancy indexing into variables, so...

```
mask = np.s_[1:3]
print(np.arange(10)[mask])
> [1 2]
```

Reply:
would be happy to test something if you have a working snippet

Reply:
Does the above not work when importing numpy?

Reply:
no, doesn't select / exclude galaxies or combine with an excplicit `structure_mask` as the original

Reply:
As in, you've tested the code and it doesn't, or it doesn't as read?

Line 776 is the line that 'includes' only galaxies of that type, 779 'excludes' galaxies, and 781 combines it with the original `structure_mask`.

Reply:
as in I've tested the code and it doesn't

Reply:
Ah I suppose the problem is that the structure mask hasn't ever been created here...